pkg/scheduler/plugins/topology/topology_plugin_job_filtering.go

```diff
@@ -0,0 +1,277 @@
+		} else {
+			break
+		}
+	}
+	// Add more to jobResourcesAllocationsRepresentors until node cannot accommodate any more pods
+	if allocateablePodsCount == len(jobAllocationMetaData.allocationTestPods) {
+		for i := allocateablePodsCount; i < len(jobAllocationMetaData.tasksToAllocate); i++ {
+			latestTestPod := jobAllocationMetaData.allocationTestPods[len(jobAllocationMetaData.allocationTestPods)-1]
+
+			iAllocationsTestPod := &pod_info.PodInfo{
+				Name:   fmt.Sprintf("%d-pods-resources", allocateablePodsCount+1),
+				ResReq: calcNextAllocationTestPodResources(latestTestPod.ResReq, jobAllocationMetaData.maxPodResources),
+			}
+			jobAllocationMetaData.allocationTestPods = append(jobAllocationMetaData.allocationTestPods, iAllocationsTestPod)
+			if node.IsTaskAllocatable(iAllocationsTestPod) {
+				allocateablePodsCount++
+			} else {
+				break
+			}
+		}
+	}
+	return allocateablePodsCount
+}
+
+func calcNextAllocationTestPodResources(previousTestResources, maxPodResources *resource_info.ResourceRequirements) *resource_info.ResourceRequirements {
+	nPlus1Resources := previousTestResources.Clone()
+	nPlus1Resources.BaseResource.Add(&maxPodResources.BaseResource)
+	if len(nPlus1Resources.GpuResourceRequirement.MigResources()) > 0 {
+		for migResource, quant := range maxPodResources.GpuResourceRequirement.MigResources() {
+			nPlus1Resources.GpuResourceRequirement.MigResources()[migResource] += quant
+		}
+	} else {
+		updatedGpuResource := resource_info.NewGpuResourceRequirementWithMultiFraction(
+			nPlus1Resources.GetNumOfGpuDevices(),
+			nPlus1Resources.GpuFractionalPortion(),
+			nPlus1Resources.GpuMemory())
+		nPlus1Resources.GpuResourceRequirement = *updatedGpuResource
+	}
+	return nPlus1Resources
+}
+
+func (t *topologyPlugin) getBestjobAllocateableDomains(job *podgroup_info.PodGroupInfo, topologyTree *TopologyInfo) ([]*TopologyDomainInfo, error) {
+	relevantLevels, err := t.calculateRelevantDomainLevels(job, topologyTree.Name, topologyTree)
+	if err != nil {
+		return nil, err
+	}
+	taskToAllocateCount := len(podgroup_info.GetTasksToAllocate(job, t.taskOrderFunc, true))
+
+	maxDepthDomains := []*TopologyDomainInfo{}
+	for _, level := range relevantLevels {
+		for _, domain := range topologyTree.DomainsByLevel[level] {
+			if domain.AllocatablePods < taskToAllocateCount { // Filter domains that cannot allocate the job
+				continue
+			}
+
+			maxDepthDomains = append(maxDepthDomains, domain)
+		}
+		if len(maxDepthDomains) > 0 {
+			break
+		}
+	}
+
+	if job.PodGroup.Spec.TopologyConstraint.PreferredTopologyLevel != "" &&
+		maxDepthDomains[0].Level != job.PodGroup.Spec.TopologyConstraint.PreferredTopologyLevel {
+		// If Preferred is defined and we couldn't find a domain on the prefered level,
+		// return a children subset and not a single domain
+		return t.improveChoiceForPreference(maxDepthDomains, job)
+	}
+
+	// For stage 1, return a single domain
+	return []*TopologyDomainInfo{maxDepthDomains[0]}, nil
```

Comment:
Wild idea, we could try each cycle another topology by using the increasing counter of scheduling attempts in the podgroup info (modulo it by the number of eligible domains or something)